DSDEGP-1576: get latest document api

pshapiro4broad opened the pull request:
### Description

DSDEGP-1576: Add API to get the most recent document given an elasticsearch index

----

### Checklist (never delete this)

Never delete this, it is our record that procedure was followed. If you find that for whatever reason one of the checklist points doesn't apply to your PR, you can leave it unchecked but please add an explanation below.

- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [x] **Submitter**: Ensure that you have added or updated tests
- [x] **Submitter**: If you're adding/switching libraries or otherwise changing the design, update the [design documentation](https://broadinstitute.atlassian.net/wiki/pages/viewpage.action?pageId=114531509).
- [x] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * Add notes on what you've tested
* Review cycle:
  * Team may comment on PR at will
  * **Reviewer assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to reviewer** for further feedback
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Move the JIRA issue to QA once this checklist is completed

pshapiro4broad committed:
DSDEGP-1370: add test to cover more code in HttpElasticSearchDAOSpec

- test updateMetadata(), queryMetaData()

pshapiro4broad committed:
Merge branch 'develop' into ps_DSDEGP-1576_get_latest_id_api

pshapiro4broad committed:
DSDEGP-1370: add test for sorted query API, add query API

pshapiro4broad committed:
DSDEGP-1370: update SearchDAO to return Document instead of MappedOutput; the caller must do the mapping now

- also updated DAO to accept QueryDefinition directly, so the caller must now map query to QueryDefinition

pshapiro4broad committed:
Merge branch 'ps_elasticsearch_output_mapping_refactor' into ps_DSDEGP-1576_get_latest_id_api

pshapiro4broad committed:
DSDEGP-1370: sorted API test is working

pshapiro4broad committed:
Merge branch 'develop' into ps_DSDEGP-1576_get_latest_id_api

# Conflicts:
#	clio-server/src/main/scala/org/broadinstitute/clio/server/dataaccess/HttpElasticsearchDAO.scala
#	clio-server/src/main/scala/org/broadinstitute/clio/server/dataaccess/SearchDAO.scala
#	clio-server/src/main/scala/org/broadinstitute/clio/server/service/SearchService.scala

pshapiro4broad committed:
Merge branch 'develop' into ps_DSDEGP-1576_get_latest_id_api

pshapiro4broad committed:
DSDEGP-1576: update to current HttpElasticSearchDAO API

pshapiro4broad committed:
DSDEGP-1576: finish 'get recent document' API

- refactor code that converts a scala name to an elasticsearch name
- add longer HTTP request timeout for elasticsearch

pshapiro4broad committed:
DSDEGP-1576: add negative test, add better error message for error case

pshapiro4broad committed:
DSDEGP-1576: javadoc updates

danxmoran commented on clio-server/src/main/scala/org/broadinstitute/clio/server/dataaccess/HttpElasticsearchDAO.scala:
```diff
@@ -30,7 +30,11 @@ class HttpElasticsearchDAO private[dataaccess] (
+    val restClient = RestClient
+      .builder(httpHosts: _*)
+      // The default timeout is 100̈ms, which is too slow for query operations.
+      .setRequestConfigCallback(_.setConnectionRequestTimeout(10000))
```
Instead of having a hard-coded value here, could we add a new key to [this config blob](https://github.com/broadinstitute/clio/blob/develop/clio-server/src/main/resources/reference.conf#L36) to parameterize the timeout?

pshapiro4broad replied:
yeah, good idea

danxmoran commented on clio-server/src/main/scala/org/broadinstitute/clio/server/dataaccess/HttpElasticsearchDAO.scala:
```diff
@@ -54,6 +58,33 @@ class HttpElasticsearchDAO private[dataaccess] (
+    * @tparam D the document type this index contains
+    * @return the most recent document for this index, if any
+    */
+  def getMostRecentDocument[D: HitReader](
```
`D <: ClioDocument` will make sure that the type we're searching over actually contains an `upsert_id` field. It'd probably make sense to add the bound to the `ElasticsearchIndex` class too.

pshapiro4broad replied:
OK

danxmoran commented on clio-server/src/main/scala/org/broadinstitute/clio/server/dataaccess/HttpElasticsearchDAO.scala:
```diff
@@ -54,6 +58,33 @@ class HttpElasticsearchDAO private[dataaccess] (
+    for {
+      searchResponse <- httpClient execute searchDefinition
+    } yield
+      searchResponse.to[D] match {
```
Nitpick: can you throw another set of braces around this whole `yield` body? They're redundant and IntelliJ complains about them but in general I think it really helps readability to wrap any multi-line statement in them.

danxmoran commented on clio-server/src/main/scala/org/broadinstitute/clio/server/dataaccess/SearchDAO.scala:
```diff
@@ -47,8 +47,9 @@ trait SearchDAO {
-    * @param index                 The index to run the query against.
-    * @tparam D                    The type of the document to query.
+    * @param index       The index to run the query against.
+
```
Accidental newline?

danxmoran commented on clio-server/src/main/scala/org/broadinstitute/clio/server/dataaccess/elasticsearch/AutoElasticsearchQueryMapper.scala:
```diff
@@ -26,9 +25,11 @@ import java.time.OffsetDateTime
+  ModelQueryOutput: ClassTag,
+  Document <: ClioDocument: ClassTag
+] private[dataaccess]
     extends ElasticsearchQueryMapper[
```
Any chance scalafmt will let you push this `extends` up to the line above?

pshapiro4broad replied:
No

danxmoran commented on clio-server/src/test/scala/org/broadinstitute/clio/server/dataaccess/HttpElasticsearchDAOSpec.scala:
```diff
@@ -105,7 +110,48 @@ class HttpElasticsearchDAOSpec
+      _ = {
+        document should be(documents.last)
+      }
+    } yield succeed
```
This can be cleaner by moving the assertion into the `yield` block:
```scala
for {
  ...
  document <- ...
} yield {
  document should be(documents.last)
}
```

danxmoran reviewed (commented):
Some suggestions, LGTM overall!

pshapiro4broad committed:
DSDEGP-1576: code review feedback

pshapiro4broad committed:
DSDEGP-1576: move request timeout setting to the config file

tbl3rd commented on clio-dataaccess-model/src/main/scala/org/broadinstitute/clio/server/dataaccess/elasticsearch/ElasticSearchUtil.scala:
```diff
@@ -0,0 +1,16 @@
+
+object ElasticSearchUtil {
+
+  /**
```
Good idea, but maybe just add this method to [Auto]ElasticsearchIndex?

pshapiro4broad replied:
It's used by `AutoElasticsearchIndex.scala`, `ElasticsearchIndex.scala`, `AutoElasticsearchQueryMapper.scala`, `ClioDocument.scala`. None of these seem more appropriate to me than another, so I put it here. But if you think one of these is best, I can move it.

pshapiro4broad committed:
Merge branch 'develop' into ps_DSDEGP-1576_get_latest_id_api

danxmoran commented on clio-server/src/main/scala/org/broadinstitute/clio/server/dataaccess/HttpElasticsearchDAO.scala:
```diff
@@ -30,7 +30,15 @@ class HttpElasticsearchDAO private[dataaccess] (
+      // The default timeout is 100̈ms, which is too slow for query operations.
+      .setRequestConfigCallback(
+        _.setConnectionRequestTimeout(
+          ClioServerConfig.Elasticsearch.httpRequestTimeout.toMicros.toInt
```
I think `setConnectionRequestTimeout` interprets its argument as millis, not micros: https://github.com/elastic/elasticsearch/blob/master/client/rest/src/main/java/org/elasticsearch/client/RestClientBuilder.java#L198

danxmoran commented on clio-dataaccess-model/src/main/scala/org/broadinstitute/clio/server/dataaccess/elasticsearch/ElasticSearchUtil.scala:
```diff
@@ -0,0 +1,16 @@
+
+import s_mach.string._
+
+object ElasticSearchUtil {
```
Nitpick: We've been using `Elasticsearch` elsewhere.

danxmoran reviewed (approved):
Some final nitpicks, but overall :+1:

tbl3rd commented on clio-server/src/main/scala/org/broadinstitute/clio/server/dataaccess/HttpElasticsearchDAO.scala:
```diff
@@ -54,6 +62,34 @@ class HttpElasticsearchDAO private[dataaccess] (
+      }
+    }
+  }
+
```
Seems like the only alternatives are there is 1 most recent document in index, or there is no document in index.
So total should be 0 on failure, though it would be enlightening to know its value, so maybe ...
s"Found ${searchResponse.hits.total} documents in ${index.indexName}."

pshapiro4broad replied:
Isn't that what the code was already doing, reporting the index name and the total hits? Or is it the wording that doesn't seem OK?

Getting back >1 documents is unlikely but not impossible as we don't have a unique key constraint on the upsert ID field, AFAIK.

pshapiro4broad committed:
DSDEGP-1576: more CR feedback

- rename classes/fields to 'Elasticsearch' not 'ElasticSearch'
- use millis not micros

pshapiro4broad committed:
Merge remote-tracking branch 'origin/ps_DSDEGP-1576_get_latest_id_api' into ps_DSDEGP-1576_get_latest_id_api

tbl3rd commented on clio-server/src/test/scala/org/broadinstitute/clio/server/dataaccess/HttpElasticsearchDAOSpec.scala:
```diff
@@ -105,7 +110,45 @@ class HttpElasticsearchDAOSpec
+        _ <- httpElasticsearchDAO.createOrUpdateIndex(index)
+        _ <- httpElasticsearchDAO.getMostRecentDocument(index)
+      } yield ()
+    }
```
recoverToExceptionIf would let you test the content of the exception with an explicit assertion.
```scala
    recoverToExceptionIf[Exception] {
      for {
        _ <- httpElasticsearchDAO.createOrUpdateIndex(index)
        _ <- httpElasticsearchDAO.getMostRecentDocument(index)
      } yield ()
    } map {
      x => x.getMessage should be("...")
    }
```

pshapiro4broad committed:
DSDEGP-1576: more CR feedback

- use recovertToExceptionIf() form to test the exception message
